### DISCUSSION TITLE
Introduce `TryParse`?

### DISCUSSION THREAD

**Discussion by rjchatfield** [pointfreeco/swift-parsing#125] [Ideas]
Feature Suggestion: Wondering if you'd contemplate a variant of `Parse { }` that takes a fallback value?

I know you have suggested this approach:

```swift
OneOf {
  Parse {
    Int.parser()
    ","
    Prefix { $0 != "," }.map(String.init)
    ","
    Bool.parser()
  }
  Always(fallbackValue)
}
```

In my own project I made a type called `TryParse` that is a helper wrapper of this.

Usage:

```swift
TryParse {
  Int.parser()
  ","
  Prefix { $0 != "," }.map(String.init)
  ","
  Bool.parser()
} catch: {
  fallbackValue
}
```

Implentation:

```swift
public struct TryParse<FirstTry>: Parser where FirstTry: Parser {
  public let parser: OneOf<Parsers.OneOf2<FirstTry, Always<Input, Output>>>

  @inlinable
  public init(
    @ParserBuilder _ tryBlock: () -> FirstTry,
    `catch` catchValue: () -> FirstTry.Output
  ) {
    parser = OneOf {
      tryBlock()
      Always<FirstTry.Input, FirstTry.Output>(catchValue())
    }
  }
  
  @inlinable
  public func parse(_ input: inout FirstTry.Input) -> FirstTry.Output {
    parser.parse(&input)!
  }
}
```

I realise I'm missing the convenience of `Parse(User.init) { }`. I respect any suggestions or bikeshedding of names.

**Comment by iampatbrown** [comment]
@randomeizer has had some cool parser suggestions as well. Have you seen [hsdoc/parse-throws](https://github.com/randomeizer/hsdoc/tree/parse-throws) which includes a [Try parser](https://github.com/randomeizer/hsdoc/blob/parse-throws/Sources/HSDocKit/Parser/Utils/Try.swift)? You two might be able to share some ideas.

**Reply by rjchatfield** [reply]
That's great! Thanks for sharing.

**Reply by randomeizer** [reply]
Yeah, I'm still baking mine, based on how the library I'm using it in develops. But something like `Try` or `TryParse` (yeah, not sure about the name exactly either) would definitely be helpful.

At the moment my library has some similar parsers which feel like they have overlap, so probably should be boiled down a little.

**Reply by stephencelis** [reply]
@randomeizer As a side note it's been very cool watching your project develop. Thanks for contributing back discussions about your experience 😄 (and PRs!)

**Reply by randomeizer** [reply]
@stephencelis No worries! I'm just glad you guys are revisiting this at the moment, since the new developments are definitely helpful. Just before you mentioned the impending `throws` branch I had started figuring out how to make returning `Result<Output,Error>` values more ergonomic to get better error output, since I didn't really want to fork it to add `throws` myself, haha.

**Comment by tgrapperon** [comment]
This is indeed a great idea. There is a `ReplaceError` parser with the same functionality in the "throw" branches:
https://github.com/pointfreeco/swift-parsing/blob/pb-throws-ep1/Sources/Parsing/Parsers/ReplaceError.swift
It accepts a `defaultValue` directly, where `Try` from @randomeizer uses a closure with the error as argument. The later seems more flexible, and `ReplaceError` should probably use a closure like this (with a dedicated init for constants) if there are no deep consequences down the line regarding printing.

**Reply by stephencelis** [reply]
`replaceError(with:)` landed in the [latest release](https://github.com/pointfreeco/swift-parsing/blob/0.7.0/Sources/Parsing/Parsers/ReplaceError.swift) (and is [on `main`](https://github.com/pointfreeco/swift-parsing/blob/main/Sources/Parsing/Parsers/ReplaceError.swift)), so we believe it can be used to capture the same functionality as the proposed `TryParse`.

We're definitely down to chat about the ergonomics of each! It'd be nice to have some overall style guide of where we should prefer parser builders vs. parser modifiers (though even in SwiftUI the distinction can feel muddy).

**Reply by tgrapperon** [reply]
@stephencelis Would there be interest in some `ParserModifier` protocol similar to `SwiftUI`, that is where the modifier doesn't depend on the generic type of the source parser?
From my experiments, this is a tricky problem, but it could maybe open the way to some interesting optimization routes.

**Reply by stephencelis** [reply]
@tgrapperon Sounds interesting! We haven't really explored what kinds of things it would unlock, but would definitely be interested in seeing what you find!

**Reply by randomeizer** [reply]
@tgrapperon What would be some use cases for a `ParserModifier`?

**Reply by tgrapperon** [reply]
@randomeizer There are some parsers in the library that are merely wrapping one parser to provide some additional derived behavior (`ReplaceError` being one of them, but also `Map`, `Filter`, etc.). In the process of wrapping, the source parser is "phagocyted" by the modifier and the new parser is mostly opaque for later modifiers. Once defined, the chain of modified parsers is set in stone, and it is very hard to derive new behaviors from the sequences of modifiers.

If modifiers are kept separated from the modified parser, we would get a well defined branch of `ModifiedParser<Parser, Modifier>`'s. Once we can see the resulting parser as a unique core parser modified by a sequence of modifiers, we can hope performing simplifications on the sequence of modifiers itself, resulting on the same effective parser, but with an optimized parsing route.

One example I would like to try to tackle with this is the problem defined in #77: Make a preconfigured unbounded `Many` fails prematurely if we only want at most 3 elements when it can parse thousands for example. The objective would be to define a kind of `ParserModifier` algebra and rewrite the chain of modifiers in some "canonical" order according to various commutation rules. If we're in a situation where `AtMost` can be brought next to `Many`, we can generate a new `Many` with an adapted upper bound when we request to build the modified parser.

I don't have yet the exact way this could happen, and this is possibly impossible to achieve, but I would like to try to see if we can find other interesting things along the way. There are surely some benefits in identifying parsers and modifiers.
Or not at all, but that would be a fun experiment!

**Reply by randomeizer** [reply]
So, you would have signature something like `ModifiedParser<Many, AtMost>`?

**Reply by tgrapperon** [reply]
Likely. But I need to tackle the typing issues first. `ViewModifier` is a protocol that comes with a `typealias`, which is not common.
The `ModifiedParser` is the parser that builds the modified parsing element, that is, that runs the `func parser(upstream: Upstream) -> some Parser` from `Modifier`, but `Upstream` is a `typealias` from `ViewModifier`, and the function is not generic over it. In SwiftUI, it aliases a type that is private to the `ViewBuilder`. How they can wrap things without erasing the types too much is still nebulous.  That's the tricky part. 

I sincerely don't know how far it can go yet, nor how much it can affect performance and other metrics. Parsers are absolutely focused on performance, but at worse, the experiment can provide ideas for the [protocolized version of TCA](https://github.com/pointfreeco/swift-composable-architecture/discussions/832) which is evolving in the same space.

**Reply by randomeizer** [reply]
Just skimming through the public API for ViewBuilder/ViewModifier, it seems like the `Content` is being mapped to a `View` via the `@ViewBuilder` annotation on `ViewModifier.body(content:)`, but I'm still wrapping my head around parts of Swift's generics implementation TBH.

Do you have an exploratory branch up somewhere for this yet?

**Reply by tgrapperon** [reply]
@randomeizer Not yet. I haven't got the time to work on this yet.

**Reply by randomeizer** [reply]
Ok, I'll keep an eye out for it.

**Comment by stephencelis** [comment]
> Feature Suggestion: Wondering if you'd contemplate a variant of `Parse { }` that takes a fallback value?

We actually added [a `replaceError(with:)` parser modifier](https://github.com/pointfreeco/swift-parsing/blob/0.7.0/Sources/Parsing/Parsers/ReplaceError.swift) in this release that achieves just that! I'll update the release notes accordingly.

_[Edit: [release notes updated](https://github.com/pointfreeco/swift-parsing/releases/tag/0.7.0) with this and a few other changes/additions, thanks!]_

> I know you have suggested this approach:
> 
> ```swift
> OneOf {
>   Parse {
>     Int.parser()
>     ","
>     Prefix { $0 != "," }.map(String.init)
>     ","
>     Bool.parser()
>   }
>   Always(fallbackValue)
> }
> ```

We've now specifically added documentation advising _against_ this approach, now that we use rethrowing parsers:

- https://github.com/pointfreeco/swift-parsing/blob/9145db33bd90079eb523498a5362e3526e267f20/Sources/Parsing/Parsers/Always.swift#L61-L63
- https://github.com/pointfreeco/swift-parsing/blob/9145db33bd90079eb523498a5362e3526e267f20/Sources/Parsing/Parsers/OneOf.swift#L37-L53

> In my own project I made a type called `TryParse` that is a helper wrapper of this.
> 
> Usage:
> 
> ```swift
> TryParse {
>   Int.parser()
>   ","
>   Prefix { $0 != "," }.map(String.init)
>   ","
>   Bool.parser()
> } catch: {
>   fallbackValue
> }
> ```

This is neat way of approaching things with builders! As I also replied [here](https://github.com/pointfreeco/swift-parsing/discussions/125#discussioncomment-2181086), we're definitely open to tweaking the ergonomics here, but it'd be a good idea to figure out a style guide for when to use parser builders over parser modifiers.

**Reply by randomeizer** [reply]
In my own `Try` implementation I was have also tossed up whether to make it a top-level Parser or a modifier. One thing I like about it being a Parser is that it stands out, which I think is probably a good thing given that it's changing error flow control quite substantially. That said, `replaceError` is simpler, and basically works like `map`.

In my case, my `Try {} catch: {err in }` lets you either return a compatible Output or throw any error, which is handy for making error messages more context-aware. For example:

```swift
let tenChars = Try {
    Parse(10) { !$0.isNewline }
} catch: { error in
    throw MyError.expected("exactly 10 non-newline characters")
}
```

This could equally be implemented as a modifier, which would look like:

```swift
let tenChars = Parse(10) { !$0.isNewline }.catchError { error in
     throw MyError.expected("exactly 10 non-newline characters")
}
```

I actually kinda prefer the Parser version, in that the parser and the error handling are at the same indentation level (in my formatting anyway), whereas the `.catch {}` version feels like more of an afterthought, and I feel would be more prone to misreading exactly what part of the parser is getting caught.

**Reply by stephencelis** [reply]
@randomeizer One consideration with making `Try`'s `catch` block throwing is the resulting parser will still be throwing, whereas `replaceError(with:)` always turns a throwing parser into a non-throwing one.

`@rethrows` doesn't seem to work for throwing closures held as properties like the one you invoke in `catch`, probably because they are dynamic, vs. a conformance, which is static. One of the primary use cases of `replaceError(with:)` is if you use it liberally, you can cook up with a parser that statically cannot throw, which is valuable for parsers that do their own domain-specific error handling, like language parsers, where parsing itself should never fail.

**Reply by randomeizer** [reply]
@stephencelis Yeah, that's true.
